lib/demo_web/controllers/user_reset_password_controller.ex

```diff
@@ -0,0 +1,59 @@
+defmodule DemoWeb.UserResetPasswordController do
+  use DemoWeb, :controller
+
+  alias Demo.Accounts
+
+  plug :get_user_by_reset_password_token when action in [:edit, :update]
+
+  def new(conn, _params) do
+    render(conn, "new.html")
+  end
+
+  def create(conn, %{"user" => %{"email" => email}}) do
+    if user = Accounts.get_user_by_email(email) do
+      Accounts.deliver_user_reset_password_instructions(
+        user,
+        &Routes.user_reset_password_url(conn, :edit, &1)
+      )
+    end
+
+    # Regardless of the outcome, show an impartial success/error message.
```

Comment:
Why show a vague message and not an explicit error message if the email isn't in the system?  I've seen this pattern used but I've never understood why.  If an attacker is trying to build a list of email addresses they can get that information by probing the sign up form.

Reply:
@gshaw yes, if you want to prevent user (email address) enumeration, you need to make sure that the sign up form (and the reset password request form) does not leak that information.

Reply:
@riverrun agreed.  My question is: Since sign up does leak emails why take the UX hit on password reset?

Reply:
@gshaw Because you could make the sign up work in this way that it would not report an error if you are using an already existing email address and just prompt you to confirm your registration using the e-mail that was sent to you.
I abstract here from judging the UX of such a behavior.

Reply:
@szymon-jez thanks for the requirements link. 👏  I'm going to assume that he is taking the stance that the end product is getting customized by all apps and he is providing a safe default with the understanding that if you want to prevent enumeration attacks you still have some work to do but whenever possible this code will help you out.  At the same time it allows people who know they can't prevent enumeration attacks since they are using email as a username to make some simple changes to improve UX on things like the password reset form.

Ultimately it gets back to the point that auth is same but different for all apps. This approach gets you 80% but you are expected to tune it to competition. I think it's a great strategy and learning tool and I could see others variations of this to show alternative ideas.